File: Memory Optimization/sol2_gemini.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

void solve()
{
    int n;
    cin >> n;
    vector<pair<int, int>> items(n);
    for (int i = 0; i < n; i++)
    {
        int x;
        cin >> x;
        int k = 0;
        while (x % 2 == 0)
        {
            x /= 2;
            k++;
        }
        items[i] = {x, k};
    }
    sort(items.begin(), items.end());

    long long ans = 0;
    int current_odd = -1;
    int current_max_k = 0;

    for (int i = 0; i < n; i++)
    {
        if (items[i].first != current_odd)
        {
            ans += current_max_k;
            current_odd = items[i].first;
            current_max_k = items[i].second;
        }
        else
        {
            current_max_k = max(current_max_k, items[i].second);
        }
    }
    ans += current_max_k;
    cout << ans << "\n";
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int t;
    if (cin >> t)
    {
        while (t--)
            solve();
    }
    return 0;
}

File: Memory Optimization/sol2_gemini.java
import java.io.*;
import java.util.*;

public class Main {
    static class FastScanner {
        InputStream in;
        byte[] buffer = new byte[1 << 16];
        int head = 0, tail = 0;

        public FastScanner(InputStream in) {
            this.in = in;
        }

        int read() throws IOException {
            if (head >= tail) {
                head = 0;
                tail = in.read(buffer, 0, buffer.length);
                if (tail <= 0) return -1;
            }
            return buffer[head++];
        }

        int nextInt() throws IOException {
            int c = read();
            while (c <= 32) {
                if (c == -1) return -1;
                c = read();
            }
            int res = 0;
            while (c > 32) {
                res = res * 10 + c - '0';
                c = read();
            }
            return res;
        }
    }

    public static void main(String[] args) throws IOException {
        FastScanner sc = new FastScanner(System.in);
        int T = sc.nextInt();
        if (T == -1) return;
        StringBuilder out = new StringBuilder();
        
        for (int t = 0; t < T; t++) {
            int N = sc.nextInt();
            // Encode (odd_part, exponent) into a primitive long to save memory/time
            long[] items = new long[N];
            
            for (int i = 0; i < N; i++) {
                int x = sc.nextInt();
                int k = Integer.numberOfTrailingZeros(x);
                int odd = x >> k;
                items[i] = ((long) odd << 32) | (k & 0xFFFFFFFFL);
            }
            
            Arrays.sort(items);
            
            long ans = 0;
            int currentOdd = -1;
            int currentMaxK = 0;
            
            for (int i = 0; i < N; i++) {
                int odd = (int) (items[i] >> 32);
                int k = (int) items[i];
                
                if (odd != currentOdd) {
                    ans += currentMaxK;
                    currentOdd = odd;
                    currentMaxK = k;
                } else {
                    if (k > currentMaxK) {
                        currentMaxK = k;
                    }
                }
            }
            ans += currentMaxK;
            out.append(ans).append("\n");
        }
        System.out.print(out);
    }
}